File: problem14.rb
$collatzCache = {}

def collatzLength(num)
  if num == 1
    return 1
  elsif $collatzCache[num]
    return $collatzCache[num]
  else
    if num.even?
      newNum = num / 2
    else
      newNum = num * 3 + 1
    end

    $collatzCache[num] = collatzLength(newNum) + 1
    return $collatzCache[num]
  end
end

answer = 0
maxLength = 0
(1...1000000).each do |i|
  len = collatzLength(i)
  if len > maxLength
    maxLength = len
    answer = i
  end
end

puts answer

File: javascript/problem14.js
var longestLength = 0;
var answer = 0;

var collatzSequence = function(num) {
  if ( num == 1 ) {
    return 1;
  } else if ( num % 2 == 0 ) {
    return 1 + collatzSequence( num / 2 );
  } else {
    return 1 + collatzSequence( ( 3 * num ) + 1 );
  }
}

for (var i = 1; i < 1000000; i++) {
  var collatzLength = collatzSequence( i );
  if ( collatzLength > longestLength ) {
    longestLength = collatzLength;
    answer = i;
  }
}

console.log( answer );
